sched/irq/irq_attach_thread.c

```diff
@@ -0,0 +1,214 @@
+/****************************************************************************
+ * sched/irq/irq_attach_thread.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <errno.h>
+#include <stdio.h>
+
+#include <nuttx/irq.h>
+#include <nuttx/kthread.h>
+
+#include "irq/irq.h"
+#include "sched/sched.h"
+
+/****************************************************************************
+ * Privte Types
+ ****************************************************************************/
+
+/* This is the type of the list of interrupt handlers, one for each IRQ.
+ * This type provided all of the information necessary to irq_dispatch to
+ * transfer control to interrupt handlers after the occurrence of an
+ * interrupt.
+ */
+
+struct irq_thread_info_s
+{
+  xcpt_t handler;     /* Address of the interrupt handler */
+  xcpt_t isrthread;   /* Address of the interrupt thread handler */
+  FAR void *arg;      /* The argument provided to the interrupt handler. */
+  sem_t sem;          /* irq sem used to notify irq thread */
+  pid_t threadid;     /* irq threadid */
+};
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct irq_thread_info_s g_irq_thread_vector[NR_IRQS];
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/* Default interrupt handler for threaded interrupts.
+ * Useful for oneshot interrupts.
+ */
+
+static int irq_default_handler(int irq, FAR void *regs, FAR void *arg)
+{
+  int ret = IRQ_WAKE_THREAD;
+  int ndx = IRQ_TO_NDX(irq);
+  xcpt_t vector;
+
+  if (ndx < 0)
+    {
+      return ndx;
+    }
+
+  vector = g_irq_thread_vector[ndx].handler;
+
+  if (vector)
+    {
+      ret = vector(irq, regs, regs);
```

Comment:
Are you sure you don't need any additional protection here?

Reply:
in isrthread，We need to disable the interrupts corresponding to the interrupt thread, so there's no need for protection here, and it won't be interrupted by the corresponding IRQ.

Reply:
But you will be interrupted by other ISR. If interrupt upper half does not have critical section protection, it may cause serious issues.

Reply:
if irq thread hold the critical section during execution, the real time behavior is same as executing the handler in hard irq directly.